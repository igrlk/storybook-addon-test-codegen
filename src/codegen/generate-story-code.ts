import { formatFileContent } from 'storybook/internal/common';
import { storyNameFromExport, toId } from 'storybook/internal/csf';
import { type CsfFile, printCsf } from 'storybook/internal/csf-tools';
import { type GeneratedCode, isPlay } from './interactions-to-code';
import {
	duplicateStoryWithNewName,
	parseArgs,
	removeExtraNewlines,
	updateArgsInCsfFile,
	updateImportsInCsfFile,
	updatePlayInCsfFile,
	updateTestsInCsfFile,
} from './save-story-utils';

export const generateStoryCode = async ({
	csf,
	csfId,
	name,
	args,
	code,
}: {
	csf: CsfFile;
	csfId: string;
	name: string;
	args: string;
	code: GeneratedCode;
}) => {
	const parsed = csf.parse();
	const stories = Object.entries(parsed._stories);

	const [componentId, storyId] = csfId.split('--');

	const originalStoryName = stories.find(([key, value]) =>
		value.id.endsWith(`--${storyId}`),
	)?.[0];
	if (!originalStoryName) {
		throw new Error('Source story not found.');
	}

	// Format the new story name to follow Storybook naming conventions
	const newStoryName = name
		.replace(/^[^a-z]/i, '')
		.replace(/^\d+/, '')
		.replace(/[^a-z0-9-_ ]/gi, '')
		.replaceAll(/([-_ ]+[a-z0-9])/gi, (match) =>
			match.toUpperCase().replace(/[-_ ]/g, ''),
		); // from https://github.com/storybookjs/storybook/blob/1fdd2d6c675b81269125af5027e45a357c09f1fa/code/addons/controls/src/SaveStory.tsx#L122

	// Check if we're updating an existing story
	const existingStoryToUpdate =
		// Check for story.name property matching our target
		stories.find(([key, story]) => {
			if (story.name === name) {
				return true;
			}

			return false;
		}) ||
		// If no story with matching name property, look for a story by its export name
		stories.find(([key]) => key === name || storyNameFromExport(key) === name);

	const newStoryId = existingStoryToUpdate
		? toId(componentId, storyId)
		: toId(componentId, storyNameFromExport(newStoryName));

	// If we're updating an existing story, use that node, otherwise create a new one
	const node = existingStoryToUpdate
		? csf._storyExports[existingStoryToUpdate[0]]
		: (csf._storyExports[newStoryName] ??
			duplicateStoryWithNewName(parsed, originalStoryName, newStoryName));

	await updateArgsInCsfFile(node, args ? parseArgs(args) : {});

	if (isPlay(code)) {
		await updatePlayInCsfFile(
			node,
			code.play.map((line) => line.text),
		);

		await updateImportsInCsfFile(
			csf._ast,
			code.imports.map((line) => line.text),
		);
	} else {
		await updateTestsInCsfFile(
			node,
			code.parameters,
			code.tests.map((line) => line.text),
			newStoryName,
			csf._ast,
		);

		await updateImportsInCsfFile(
			csf._ast,
			code.imports.map((line) => line.text),
		);
	}

	return {
		storyCode: await formatFileContent(
			'.',
			removeExtraNewlines(printCsf(csf).code, newStoryName),
		),
		newStoryId,
	};
};
