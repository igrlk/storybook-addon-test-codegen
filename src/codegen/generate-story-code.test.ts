import { loadCsf } from 'storybook/internal/csf-tools';
import { describe, expect, it } from 'vitest';
import { generateStoryCode } from './generate-story-code';
import type { GeneratedCode } from './interactions-to-code';

const TEST_CASES = [
	[
		'New story and new imports',
		{
			story: `import type { Meta, StoryObj } from '@storybook/react';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {};`,
			code: {
				imports: ["import { userEvent, within } from '@storybook/test';"],
				play: [
					'play: async ({ canvasElement }) => {',
					'\tconst canvas = within(canvasElement.ownerDocument.body);',
					"\tawait userEvent.click(await canvas.findByRole('button'));",
					'}',
				],
			},
			result: `import { userEvent, within } from '@storybook/test';
import type { Meta, StoryObj } from '@storybook/react';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {};

export const NewStory: Story = {
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement.ownerDocument.body);
        await userEvent.click(await canvas.findByRole('button'));
    }
};`,
		},
	],

	[
		'New story with spaces in name and some existing imports',
		{
			story: `import { userEvent } from '@storybook/test';
import type { Meta, StoryObj } from '@storybook/react';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {};`,
			code: {
				imports: ["import { userEvent, within } from '@storybook/test';"],
				play: [
					'play: async ({ canvasElement }) => {',
					'\tconst canvas = within(canvasElement.ownerDocument.body);',
					"\tawait userEvent.click(await canvas.findByRole('button'));",
					'}',
				],
			},
			result: `import { userEvent, within } from '@storybook/test';
import type { Meta, StoryObj } from '@storybook/react';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {};

export const NewStory: Story = {
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement.ownerDocument.body);
        await userEvent.click(await canvas.findByRole('button'));
    }
};`,
		},
	],

	[
		'New story and all existing imports',
		{
			story: `import { userEvent, within } from '@storybook/test';
import type { Meta, StoryObj } from '@storybook/react';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {};`,
			code: {
				imports: ["import { userEvent, within } from '@storybook/test';"],
				play: [
					'play: async ({ canvasElement }) => {',
					'\tconst canvas = within(canvasElement.ownerDocument.body);',
					"\tawait userEvent.click(await canvas.findByRole('button'));",
					'}',
				],
			},
			result: `import { userEvent, within } from '@storybook/test';
import type { Meta, StoryObj } from '@storybook/react';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {};

export const NewStory: Story = {
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement.ownerDocument.body);
        await userEvent.click(await canvas.findByRole('button'));
    }
};`,
		},
	],

	[
		'Existing story',
		{
			storyName: 'Default',
			story: `import type { Meta, StoryObj } from '@storybook/react';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {};`,
			code: {
				imports: ["import { userEvent, within } from '@storybook/test';"],
				play: [
					'play: async ({ canvasElement }) => {',
					'\tconst canvas = within(canvasElement.ownerDocument.body);',
					"\tawait userEvent.click(await canvas.findByRole('button'));",
					'}',
				],
			},
			result: `import { userEvent, within } from '@storybook/test';
import type { Meta, StoryObj } from '@storybook/react';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement.ownerDocument.body);
        await userEvent.click(await canvas.findByRole('button'));
    }
};`,
		},
	],

	[
		'Pure javascript',
		{
			storyName: '123New Story',
			story: `import { Component } from './Component';

export default {
    component: Component
};

export const Default = {};`,
			code: {
				imports: ["import { userEvent, within } from '@storybook/test';"],
				play: [
					'play: async ({ canvasElement }) => {',
					'\tconst canvas = within(canvasElement.ownerDocument.body);',
					"\tawait userEvent.click(await canvas.findByRole('button'));",
					'}',
				],
			},
			result: `import { userEvent, within } from '@storybook/test';
import { Component } from './Component';

export default {
    component: Component
};

export const Default = {};

export const NewStory = {
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement.ownerDocument.body);
        await userEvent.click(await canvas.findByRole('button'));
    }
};`,
		},
	],

	[
		'New story with args',
		{
			storyName: 'NewStory',
			args: JSON.stringify({
				hello: {
					world: {
						foo: 'bar',
						baz: [1, 2, { boo: 'bee' }],
					},
				},
			}),
			story: `import type { Meta, StoryObj } from '@storybook/react';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {};`,
			code: {
				imports: ["import { userEvent, within } from '@storybook/test';"],
				play: [
					'play: async ({ canvasElement }) => {',
					'\tconst canvas = within(canvasElement.ownerDocument.body);',
					"\tawait userEvent.click(await canvas.findByRole('button'));",
					'}',
				],
			},
			result: `import { userEvent, within } from '@storybook/test';
import type { Meta, StoryObj } from '@storybook/react';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {};

export const NewStory: Story = {
    args: {
        hello: {
            "world": {
                "foo": "bar",

                "baz": [1, 2, {
                    "boo": "bee"
                }]
            }
        }
    },

    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement.ownerDocument.body);
        await userEvent.click(await canvas.findByRole('button'));
    }
};`,
		},
	],

	[
		'Existing story with args',
		{
			storyName: 'Default',
			args: JSON.stringify({
				hello: {
					world: {
						foo: 'bar',
						baz: [1, 2, { boo: 'bee' }],
					},
				},
			}),
			story: `import type { Meta, StoryObj } from '@storybook/react';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {
    args: {
        existing: "existing value"
    }
};`,
			code: {
				imports: ["import { userEvent, within } from '@storybook/test';"],
				play: [
					'play: async ({ canvasElement }) => {',
					'\tconst canvas = within(canvasElement.ownerDocument.body);',
					"\tawait userEvent.click(await canvas.findByRole('button'));",
					'}',
				],
			},
			result: `import { userEvent, within } from '@storybook/test';
import type { Meta, StoryObj } from '@storybook/react';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {
    args: {
        existing: "existing value",

        hello: {
            "world": {
                "foo": "bar",

                "baz": [1, 2, {
                    "boo": "bee"
                }]
            }
        }
    },

    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement.ownerDocument.body);
        await userEvent.click(await canvas.findByRole('button'));
    }
};`,
		},
	],
] satisfies [
	string,
	{
		storyName?: string;
		args?: string;
		story: string;
		code: GeneratedCode;
		result: string;
	},
][];

describe('generate-story-code', () => {
	it.each(TEST_CASES)('%s', async (_, params) => {
		expect(
			await generateStoryCode({
				code: params.code,
				name: 'storyName' in params ? params.storyName : 'NewStory',
				args: 'args' in params ? params.args : '{}',
				csfId: 'form--default',
				csf: loadCsf(params.story, { makeTitle: () => 'story' }),
			}),
		).toBe(params.result);
	});
});
