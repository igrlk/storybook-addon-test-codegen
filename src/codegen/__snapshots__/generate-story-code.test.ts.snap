// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateStoryCode > Play function syntax > Existing story 1`] = `
"import { userEvent, within } from 'storybook/test';
import type { Meta, StoryObj } from '@storybook/react-vite';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement.ownerDocument.body);
        await userEvent.click(await canvas.findByRole('button'));
    }
};"
`;

exports[`generateStoryCode > Play function syntax > Existing story with args 1`] = `
"import { userEvent, within } from 'storybook/test';
import type { Meta, StoryObj } from '@storybook/react-vite';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {
    args: {
        existing: "existing value",
        someFunction: () => 42,

        hello: {
            "world": {
                "foo": "bar",

                "baz": [1, 2, {
                    "boo": "bee"
                }]
            }
        }
    },

    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement.ownerDocument.body);
        await userEvent.click(await canvas.findByRole('button'));
    }
};"
`;

exports[`generateStoryCode > Play function syntax > Existing story with custom name 1`] = `
"import { userEvent, within } from 'storybook/test';
import type { Meta, StoryObj } from '@storybook/react-vite';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const CustomName: Story = {
    name: "My cool name",

    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement.ownerDocument.body);
        await userEvent.click(await canvas.findByRole('button'));
    }
};"
`;

exports[`generateStoryCode > Play function syntax > New story and all existing imports 1`] = `
"import { userEvent, within } from 'storybook/test';
import type { Meta, StoryObj } from '@storybook/react-vite';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {};

export const NewStory: Story = {
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement.ownerDocument.body);
        await userEvent.click(await canvas.findByRole('button'));
    }
};"
`;

exports[`generateStoryCode > Play function syntax > New story and new imports 1`] = `
"import { userEvent, within } from 'storybook/test';
import type { Meta, StoryObj } from '@storybook/react-vite';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {};

export const NewStory: Story = {
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement.ownerDocument.body);
        await userEvent.click(await canvas.findByRole('button'));
    }
};"
`;

exports[`generateStoryCode > Play function syntax > New story with args 1`] = `
"import { userEvent, within } from 'storybook/test';
import type { Meta, StoryObj } from '@storybook/react-vite';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {};

export const NewStory: Story = {
    args: {
        hello: {
            "world": {
                "foo": "bar",

                "baz": [1, 2, {
                    "boo": "bee"
                }]
            }
        }
    },

    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement.ownerDocument.body);
        await userEvent.click(await canvas.findByRole('button'));
    }
};"
`;

exports[`generateStoryCode > Play function syntax > New story with spaces in name and some existing imports 1`] = `
"import { userEvent, within } from 'storybook/test';
import type { Meta, StoryObj } from '@storybook/react-vite';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {};

export const NewStory: Story = {
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement.ownerDocument.body);
        await userEvent.click(await canvas.findByRole('button'));
    }
};"
`;

exports[`generateStoryCode > Play function syntax > New story without copying name 1`] = `
"import { userEvent, within } from 'storybook/test';
import type { Meta, StoryObj } from '@storybook/react-vite';
import { Component } from './Component';

const meta: Meta<typeof Component> = {
    component: Component
};
export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {
    name: "Default"
};

export const NewStory: Story = {
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement.ownerDocument.body);
        await userEvent.click(await canvas.findByRole('button'));
    }
};"
`;

exports[`generateStoryCode > Play function syntax > Pure javascript 1`] = `
"import { userEvent, within } from 'storybook/test';
import { Component } from './Component';

export default {
    component: Component
};

export const Default = {};

export const NewStory12 = {
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement.ownerDocument.body);
        await userEvent.click(await canvas.findByRole('button'));
    }
};"
`;

exports[`generateStoryCode > Test syntax > Existing story with args and test syntax 1`] = `
"import preview from '../.storybook/preview';
import { Component } from './Component';

const meta = preview.meta({
  component: Component
});

export const Default = meta.story({
    args: {
        existing: "existing value",
        someFunction: () => 42,

        hello: {
            "world": {
                "foo": "bar",

                "baz": [1, 2, {
                    "boo": "bee"
                }]
            }
        }
    }
});

Default.test("typed-test-name", async (
    {
        canvas,
        userEvent
    }
) => {
    const body = canvas.ownerDocument.body;
    await userEvent.click(await canvas.findByRole("button"));
});"
`;

exports[`generateStoryCode > Test syntax > Existing story with test syntax 1`] = `
"import preview from '../.storybook/preview';
import { Component } from './Component';

const meta = preview.meta({
  component: Component
});

export const Default = meta.story({});

Default.test("typed-test-name", async (
  {
    canvas,
    userEvent
  }
) => {
  const body = canvas.ownerDocument.body;
  await userEvent.click(await canvas.findByRole("button"));
});"
`;

exports[`generateStoryCode > Test syntax > New story and all existing imports with test syntax 1`] = `
"import preview from '../.storybook/preview';
import { Component } from './Component';

const meta = preview.meta({
  component: Component
});

export const Default = meta.story({});
export const NewStory = meta.story({});

NewStory.test("typed-test-name", async (
  {
    canvas,
    userEvent
  }
) => {
  const body = canvas.ownerDocument.body;
  await userEvent.click(await canvas.findByRole("button"));
});"
`;

exports[`generateStoryCode > Test syntax > New story and new imports with test syntax 1`] = `
"import { waitFor, expect } from 'storybook/test';
import preview from '../.storybook/preview';
import { Component } from './Component';

const meta = preview.meta({
  component: Component
});

export const Default = meta.story({});
export const NewStory = meta.story({});

NewStory.test("typed-test-name", async (
  {
    canvas,
    userEvent
  }
) => {
  const body = canvas.ownerDocument.body;
  await userEvent.click(await canvas.findByRole("button"));
  await waitFor(() => expect(canvas.queryByRole("button")).toBeInTheDocument());
});"
`;

exports[`generateStoryCode > Test syntax > New story with args and test syntax 1`] = `
"import preview from '../.storybook/preview';
import { Component } from './Component';

const meta = preview.meta({
  component: Component
});

export const Default = meta.story({});

export const NewStory = meta.story({
  args: {
    hello: {
      "world": {
        "foo": "bar",

        "baz": [1, 2, {
          "boo": "bee"
        }]
      }
    }
  }
});

NewStory.test("typed-test-name", async (
  {
    canvas,
    userEvent
  }
) => {
  const body = canvas.ownerDocument.body;
  await userEvent.click(await canvas.findByRole("button"));
});"
`;

exports[`generateStoryCode > Test syntax > New story with spaces in name and some existing imports with test syntax 1`] = `
"import preview from '../.storybook/preview';
import { Component } from './Component';

const meta = preview.meta({
  component: Component
});

export const Default = meta.story({});
export const NewStory = meta.story({});

NewStory.test("typed-test-name", async (
  {
    canvas,
    userEvent
  }
) => {
  const body = canvas.ownerDocument.body;
  await userEvent.click(await canvas.findByRole("button"));
});"
`;
